src/java/org/apache/cassandra/index/sai/utils/AbstractIterator.java

```diff
@@ -0,0 +1,155 @@
+/*
+ * Copyright (C) 2007 The Guava Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.index.sai.utils;
+
+import java.util.NoSuchElementException;
+
+import com.google.common.collect.PeekingIterator;
+
+import static com.google.common.base.Preconditions.checkState;
+
+// This is fork of the Guava AbstractIterator, the only difference
+// is that state & next variables are now protected, this was required
+// for SkippableIterator.skipTo(..) to void all previous state.
+public abstract class AbstractIterator<T> implements PeekingIterator<T>
```

Comment:
This is a verbatim copy of `org.apache.cassandra.index.sasi.utils.AbstractIterator`. Why don't we rename it something like `GuavaAbstractIterator`, move it to `org.apache.cassandra.utils`, and remove the `sasi` version from the codebase?

(I suppose we could then later try to merge/consolidate `GuavaAbstractIterator` and `AbstractIterator`, but let's at least not explicitly duplicate things for now.)

Reply:
Agreed but I have gone with `AbstractGuavaIterator` just to keep it consistent with other abstract classes. I have sorted the javadoc on this as requested above.